src/crypto/key_wrap.cpp

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+
+#include "key_wrap.h"
+
+#include "crypto/key_pair.h"
+#include "crypto/symmetric_key.h"
+#include "openssl/symmetric_key.h"
+#include "rsa_key_pair.h"
+
+#include <cstdint>
+#include <stdexcept>
+#include <vector>
+
+namespace crypto
+{
+  // With inspiration from
+  // http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/cos01/pkcs11-curr-v2.40-cos01.html
+
+  std::vector<uint8_t> ckm_rsa_pkcs_oaep_wrap(
+    RSAPublicKeyPtr wrapping_key,
+    const std::vector<uint8_t>& unwrapped,
+    const std::vector<uint8_t>& label)
+  {
+    return wrapping_key->wrap(unwrapped, label);
+  }
+
+  std::vector<uint8_t> ckm_rsa_pkcs_oaep_wrap(
+    const Pem& wrapping_key,
+    const std::vector<uint8_t>& unwrapped,
+    const std::vector<uint8_t>& label)
+  {
+    auto pk = make_rsa_public_key(wrapping_key);
+    return ckm_rsa_pkcs_oaep_wrap(pk, unwrapped, label);
+  }
+
+  std::vector<uint8_t> ckm_rsa_pkcs_oaep_unwrap(
+    RSAKeyPairPtr wrapping_key,
+    const std::vector<uint8_t>& wrapped,
+    const std::vector<uint8_t>& label)
+  {
+    return wrapping_key->unwrap(wrapped, label);
+  }
+
+  std::vector<uint8_t> ckm_rsa_pkcs_oaep_unwrap(
+    const Pem& wrapping_key,
+    const std::vector<uint8_t>& wrapped,
+    const std::vector<uint8_t>& label)
+  {
+    auto pk = make_rsa_key_pair(wrapping_key);
+    return ckm_rsa_pkcs_oaep_unwrap(pk, wrapped, label);
+  }
+
+  std::vector<uint8_t> ckm_aes_key_wrap_pad(
+    const std::vector<uint8_t>& wrapping_key,
+    const std::vector<uint8_t>& unwrapped)
+  {
+    auto ossl = std::make_unique<KeyAesGcm_OpenSSL>(wrapping_key);
```

Comment:
Can we avoid hard-coding to OpenSSL somehow?

Reply:
No, we can't. The algorithm comes with mbedTLS too, but the Open Enclave config disables it. That's not a problem though, we have OpenSSL available now and forever and there are no knock-on effects like having to export/import PEMs to switch between the two.